packages/compass-aggregations/src/modules/pipeline-builder/text-editor.ts

```diff
@@ -0,0 +1,209 @@
+    ) ||
+    isAction<PipelineModeToggledAction>(
+      action,
+      PipelineModeActionTypes.PipelineModeToggled
+    )
+  ) {
+    return {
+      ...state,
+      pipelineText: action.pipelineText,
+      stageOperators: action.stageOperators,
+      syntaxErrors: action.syntaxErrors,
+    };
+  }
+
+  if (
+    isAction<EditorPreviewFetchAction>(
+      action,
+      EditorActionTypes.EditorPreviewFetch
+    )
+  ) {
+    return {
+      ...state,
+      previewDocs: null,
+      serverError: null,
+      loading: true,
+    };
+  }
+
+  if (
+    isAction<EditorPreviewFetchSuccessAction>(
+      action,
+      EditorActionTypes.EditorPreviewFetchSuccess
+    )
+  ) {
+    return {
+      ...state,
+      serverError: null,
+      syntaxErrors: [],
+      loading: false,
+      previewDocs: action.previewDocs,
+    };
+  }
+
+  if (
+    isAction<EditorPreviewFetchErrorAction>(
+      action,
+      EditorActionTypes.EditorPreviewFetchError
+    )
+  ) {
+    return {
+      ...state,
+      serverError: action.serverError,
+      syntaxErrors: [],
+      loading: false,
+      previewDocs: null,
+    };
+  }
+
+  return state;
+};
+
+export const loadPreviewForPipeline = (
+): PipelineBuilderThunkAction<
+  Promise<void>,
+  EditorPreviewFetchAction |
+  EditorPreviewFetchSuccessAction |
+  EditorPreviewFetchErrorAction
+> => {
+  return async (dispatch, getState, { pipelineBuilder }) => {
+    const {
+      autoPreview
+    } = getState();
+
+    const canRunPipeline = pipelineBuilder.syntaxError.length === 0;
+    if (!autoPreview || !canRunPipeline) {
+      return;
+    }
+
+    try {
+      dispatch({
+        type: EditorActionTypes.EditorPreviewFetch,
+      });
+
+      const {
+        namespace,
+        maxTimeMS,
+        collationString,
+        limit,
+        largeLimit,
+        inputDocuments
+      } = getState();
+
+      const options: PreviewOptions = {
+        maxTimeMS: maxTimeMS ?? DEFAULT_MAX_TIME_MS,
+        collation: collationString.value ?? undefined,
+        sampleSize: largeLimit ?? DEFAULT_SAMPLE_SIZE,
+        previewSize: limit ?? DEFAULT_PREVIEW_LIMIT,
+        totalDocumentCount: inputDocuments.count
+      };
+
+      const previewDocs = await pipelineBuilder.getPreviewForPipeline(
+        namespace,
+        options
+      );
+
+      dispatch({
+        type: EditorActionTypes.EditorPreviewFetchSuccess,
+        previewDocs
+      });
+    } catch (err) {
+      if (isCancelError(err)) {
+        return;
+      }
+      dispatch({
+        type: EditorActionTypes.EditorPreviewFetchError,
+        serverError: err as MongoServerError
+      });
+    }
+  };
+};
+
+export const changeEditorValue = (
+  value: string
+): PipelineBuilderThunkAction<void, EditorValueChangeAction> => {
+  return (dispatch, getState, { pipelineBuilder }) => {
+    pipelineBuilder.changeSource(value);
+    const stageOperators = getStageOperatorsFromPipelineSource(
+      getState(),
+      pipelineBuilder
+    );
```

Comment:
As we are doing this on every source change, I am starting to think that maybe it makes sense to keep the parsed pipeline and the logic for getting it on source update on the pipeline builder instance

Reply:
added it in [`2f5a100` (#3620)](https://github.com/mongodb-js/compass/pull/3620/commits/2f5a10089d3170c4050a6bcc46d2dc3c2a377985)